File: java/src/test/java/myairtable/tests/TestCaching.java
package myairtable.tests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import myairtable.Airtable;
import myairtable.AirtableQuery;
import myairtable.CacheStore;
import myairtable.PrimaryModel;
import myairtable.PrimaryTable;
import org.junit.jupiter.api.Test;

/**
 * J-port of Kotlin K9.5 — TTL cache behavior end-to-end: hits, default-off, mutation invalidation,
 * manual invalidation, TTL expiry, query-key separation. Parity with Kotlin TestCaching / Swift
 * TestCaching / Rust test_caching.
 */
class TestCaching {

  private static Airtable cachedAirtable() {
    return TestSetup.makeAirtable(60.0);
  }

  private static Airtable uncachedAirtable() {
    return TestSetup.makeAirtable(); // cacheSeconds defaults to 0
  }

  private static void bestEffortDelete(Airtable at, String recordId) {
    try {
      at.primary().delete(recordId);
    } catch (RuntimeException ignored) {
      // best-effort cleanup
    }
  }

  @Test
  void cacheHitReturnsSameRecordOnRepeatedGet() {
    Airtable at = cachedAirtable();
    String primaryKey = TestSetup.primaryKey("Cache", "Hit");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    try {
      PrimaryModel first = at.primary().get(id);
      PrimaryModel second = at.primary().get(id);
      assertEquals(first.getPrimaryKey(), second.getPrimaryKey());
      assertEquals(first.getId(), second.getId());
      assertTrue(at.client().getCache().count() > 0, "read populated the cache");
      at.primary().delete(id);
    } catch (Throwable e) {
      bestEffortDelete(at, id);
      throw e;
    }
  }

  @Test
  void cacheIsDisabledWhenCacheSecondsIsZero() {
    Airtable at = uncachedAirtable();
    assertEquals(0, at.client().getCache().count());
    String primaryKey = TestSetup.primaryKey("Cache", "Off");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    try {
      at.primary().get(id);
      // Even after a read the cache stays empty because TTL=0 is a no-op.
      assertEquals(0, at.client().getCache().count());
      at.primary().delete(id);
    } catch (Throwable e) {
      bestEffortDelete(at, id);
      throw e;
    }
  }

  @Test
  void createInvalidatesCachedReadsForTheTable() {
    Airtable at = cachedAirtable();
    String primaryKey = TestSetup.primaryKey("Cache", "MutateCreate");
    PrimaryModel a = at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String idA = a.getId();
    // Track every created id so a failing assert still cleans up all of them, not just idA.
    String idB = null;
    try {
      at.primary().get(idA); // populate cache
      assertTrue(at.client().getCache().count() > 0);

      PrimaryModel b =
          at.primary().create(PrimaryModel.builder().primaryKey(primaryKey + " B").build());
      idB = b.getId();
      assertEquals(0, at.client().getCache().count(), "create wipes this table's cache");

      at.primary().deleteAll(List.of(idA, idB));
    } catch (Throwable e) {
      bestEffortDelete(at, idA);
      if (idB != null) {
        bestEffortDelete(at, idB);
      }
      throw e;
    }
  }

  @Test
  void updateInvalidatesCachedReadsForTheTable() {
    Airtable at = cachedAirtable();
    String primaryKey = TestSetup.primaryKey("Cache", "MutateUpdate");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    try {
      at.primary().get(id);
      assertTrue(at.client().getCache().count() > 0);

      created.setSingleLineText("mutated");
      at.primary().update(created);
      assertEquals(0, at.client().getCache().count(), "update wipes this table's cache");

      at.primary().delete(id);
    } catch (Throwable e) {
      bestEffortDelete(at, id);
      throw e;
    }
  }

  @Test
  void deleteInvalidatesCachedReadsForTheTable() {
    Airtable at = cachedAirtable();
    String primaryKey = TestSetup.primaryKey("Cache", "MutateDelete");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    // The cache-populating read and its assert were previously outside any try, so a failing
    // count() assert leaked the created record. Guard the whole sequence.
    boolean deleted = false;
    try {
      at.primary().get(id);
      assertTrue(at.client().getCache().count() > 0);

      at.primary().delete(id);
      deleted = true;
      assertEquals(0, at.client().getCache().count(), "delete wipes this table's cache");
    } catch (Throwable e) {
      if (!deleted) {
        bestEffortDelete(at, id);
      }
      throw e;
    }
  }

  @Test
  void invalidateCacheDropsThisTablesCache() {
    Airtable at = cachedAirtable();
    String primaryKey = TestSetup.primaryKey("Cache", "ManualTable");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    try {
      at.primary().get(id);
      assertTrue(at.client().getCache().count() > 0);

      at.client().invalidateCache(PrimaryTable.TABLE_ID);
      assertEquals(0, at.client().getCache().count());

      at.primary().delete(id);
    } catch (Throwable e) {
      bestEffortDelete(at, id);
      throw e;
    }
  }

  @Test
  void invalidateAllCachesDropsEveryTablesCache() {
    Airtable at = cachedAirtable();
    String primaryKey = TestSetup.primaryKey("Cache", "ManualAll");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    try {
      at.primary().get(id);
      at.secondary().get(new AirtableQuery().withMaxRecords(1));
      assertTrue(at.client().getCache().count() >= 2);

      at.invalidateAllCaches();
      assertEquals(0, at.client().getCache().count());

      at.primary().delete(id);
    } catch (Throwable e) {
      bestEffortDelete(at, id);
      throw e;
    }
  }

  @Test
  void cachedEntriesExpireAfterTheConfiguredTtl() throws InterruptedException {
    Airtable at = TestSetup.makeAirtable(1.0);
    String primaryKey = TestSetup.primaryKey("Cache", "TTL");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    try {
      at.primary().get(id);
      assertTrue(at.client().getCache().count() > 0);

      Thread.sleep(1_500);
      // Distinguish genuine TTL expiry from a stale hit: simply re-reading the same payload would
      // pass whether or not expiry works (a stale entry returns the same record). Probe the cache
      // directly for the single-record entry — CacheStore.get() lazily evicts an entry past its
      // TTL and returns null, so a null here proves the pre-sleep entry actually expired rather
      // than being served stale. (Key shape mirrors AirtableClient.getRecord: "rec:" + id.)
      CacheStore.Key recordKey = new CacheStore.Key(PrimaryTable.TABLE_ID, "rec:" + id);
      assertNull(
          at.client().getCache().get(recordKey),
          "TTL-expired entry must be evicted on access, not served stale");

      // A fresh fetch still works (and re-populates the cache).
      PrimaryModel fresh = at.primary().get(id);
      assertEquals(primaryKey, fresh.getPrimaryKey());

      at.primary().delete(id);
    } catch (Throwable e) {
      bestEffortDelete(at, id);
      throw e;
    }
  }

  @Test
  void differentQueryParamsProduceDifferentCacheKeys() {
    Airtable at = cachedAirtable();
    String primaryKey = TestSetup.primaryKey("Cache", "Keys");
    PrimaryModel created =
        at.primary().create(PrimaryModel.builder().primaryKey(primaryKey).build());
    String id = created.getId();
    try {
      at.primary().get(new AirtableQuery().withMaxRecords(1));
      at.primary().get(new AirtableQuery().withMaxRecords(2));
      assertTrue(at.client().getCache().count() >= 2, "distinct queries cache separately");

      at.primary().delete(id);
    } catch (Throwable e) {
      bestEffortDelete(at, id);
      throw e;
    }
  }
}

File: kotlin/src/test/kotlin/myairtable/tests/TestCaching.kt
package myairtable.tests

import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import myairtable.AirtableQuery
import myairtable.PrimaryModel
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.seconds

/**
 * K9.5 — TTL cache behavior end-to-end: hits, default-off, mutation
 * invalidation, manual invalidation, TTL expiry, query-key separation.
 * Parity with Swift TestCaching / Rust test_caching.
 */
class TestCaching {
    private fun cachedAirtable() = TestSetup.makeAirtable(cacheSeconds = 60.0)

    private fun uncachedAirtable() = TestSetup.makeAirtable() // cacheSeconds defaults to 0

    @Test
    fun cacheHitReturnsSameRecordOnRepeatedGet() =
        runBlocking {
            val at = cachedAirtable()
            val primaryKey = TestSetup.primaryKey("Cache", "Hit")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            try {
                val first = at.primary.get(id)
                val second = at.primary.get(id)
                assertEquals(first.primaryKey, second.primaryKey)
                assertEquals(first.id, second.id)
                assertTrue(at.client.cache.count() > 0, "read populated the cache")
                at.primary.delete(id)
            } catch (e: Throwable) {
                runCatching { at.primary.delete(id) }
                throw e
            }
        }

    @Test
    fun cacheIsDisabledWhenCacheSecondsIsZero() =
        runBlocking {
            val at = uncachedAirtable()
            assertEquals(0, at.client.cache.count())
            val primaryKey = TestSetup.primaryKey("Cache", "Off")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            try {
                at.primary.get(id)
                // Even after a read the cache stays empty because TTL=0 is a no-op.
                assertEquals(0, at.client.cache.count())
                at.primary.delete(id)
            } catch (e: Throwable) {
                runCatching { at.primary.delete(id) }
                throw e
            }
        }

    @Test
    fun createInvalidatesCachedReadsForTheTable() =
        runBlocking {
            val at = cachedAirtable()
            val primaryKey = TestSetup.primaryKey("Cache", "MutateCreate")
            val a = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val idA = a.id!!
            try {
                at.primary.get(idA) // populate cache
                assertTrue(at.client.cache.count() > 0)

                val b = at.primary.create(PrimaryModel(primaryKey = "$primaryKey B"))
                assertEquals(0, at.client.cache.count(), "create wipes this table's cache")

                at.primary.delete(listOf(idA, b.id!!))
            } catch (e: Throwable) {
                runCatching { at.primary.delete(idA) }
                throw e
            }
        }

    @Test
    fun updateInvalidatesCachedReadsForTheTable() =
        runBlocking {
            val at = cachedAirtable()
            val primaryKey = TestSetup.primaryKey("Cache", "MutateUpdate")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            try {
                at.primary.get(id)
                assertTrue(at.client.cache.count() > 0)

                created.singleLineText = "mutated"
                at.primary.update(created)
                assertEquals(0, at.client.cache.count(), "update wipes this table's cache")

                at.primary.delete(id)
            } catch (e: Throwable) {
                runCatching { at.primary.delete(id) }
                throw e
            }
        }

    @Test
    fun deleteInvalidatesCachedReadsForTheTable() =
        runBlocking {
            val at = cachedAirtable()
            val primaryKey = TestSetup.primaryKey("Cache", "MutateDelete")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            at.primary.get(id)
            assertTrue(at.client.cache.count() > 0)

            at.primary.delete(id)
            assertEquals(0, at.client.cache.count(), "delete wipes this table's cache")
        }

    @Test
    fun invalidateCacheDropsThisTablesCache() =
        runBlocking {
            val at = cachedAirtable()
            val primaryKey = TestSetup.primaryKey("Cache", "ManualTable")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            try {
                at.primary.get(id)
                assertTrue(at.client.cache.count() > 0)

                at.client.invalidateCache(myairtable.PrimaryTable.TABLE_ID)
                assertEquals(0, at.client.cache.count())

                at.primary.delete(id)
            } catch (e: Throwable) {
                runCatching { at.primary.delete(id) }
                throw e
            }
        }

    @Test
    fun invalidateAllCachesDropsEveryTablesCache() =
        runBlocking {
            val at = cachedAirtable()
            val primaryKey = TestSetup.primaryKey("Cache", "ManualAll")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            try {
                at.primary.get(id)
                at.secondary.get(AirtableQuery(maxRecords = 1))
                assertTrue(at.client.cache.count() >= 2)

                at.invalidateAllCaches()
                assertEquals(0, at.client.cache.count())

                at.primary.delete(id)
            } catch (e: Throwable) {
                runCatching { at.primary.delete(id) }
                throw e
            }
        }

    @Test
    fun cachedEntriesExpireAfterTheConfiguredTtl() =
        runBlocking {
            val at = TestSetup.makeAirtable(cacheSeconds = 1.0)
            val primaryKey = TestSetup.primaryKey("Cache", "TTL")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            try {
                at.primary.get(id)
                assertTrue(at.client.cache.count() > 0)

                delay(1.5.seconds)
                // Expired entry is lazily evicted on access; a fresh fetch still works.
                val fresh = at.primary.get(id)
                assertEquals(primaryKey, fresh.primaryKey)

                at.primary.delete(id)
            } catch (e: Throwable) {
                runCatching { at.primary.delete(id) }
                throw e
            }
        }

    @Test
    fun differentQueryParamsProduceDifferentCacheKeys() =
        runBlocking {
            val at = cachedAirtable()
            val primaryKey = TestSetup.primaryKey("Cache", "Keys")
            val created = at.primary.create(PrimaryModel(primaryKey = primaryKey))
            val id = created.id!!
            try {
                at.primary.get(AirtableQuery(maxRecords = 1))
                at.primary.get(AirtableQuery(maxRecords = 2))
                assertTrue(at.client.cache.count() >= 2, "distinct queries cache separately")

                at.primary.delete(id)
            } catch (e: Throwable) {
                runCatching { at.primary.delete(id) }
                throw e
            }
        }
}
